Update ports controller to use ARP/MAC table lookup for IPv4 neighbours

eskyuu opened the pull request:
We have a lot of VRFs that share the same IPv4 address space.  As a result, the IPv4 network lookup code gives a large number of invalid neighbours.

This PR changes the port controller to use the ARP/MAC table relationship to resolve IPv4 neighbours, which results in more reliable link lookups between routers.  It may result in missing results if a device does not have ARP table information in NMS.  It will also have a reduced number of neighbours if a device does not have an ARP entry for some devices in the same subnet.

This shouldn't be needed for IPv6 since duplicate addresses shouldn't happen in IPv6 address space.

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [ ] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [ ] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

eskyuu commented:
I believe that this is removing the only use case of the ipv4_networks code.  If so, should this model be removed?

The other thought I had was to make the peer detection method a new config option, and keep the original code as the default.

laf commented:
> I believe that this is removing the only use case of the ipv4_networks code. If so, should this model be removed?

For Models I'd say no leave it (we have an API PR in now anyway that will use it).

> The other thought I had was to make the peer detection method a new config option, and keep the original code as the default.

I definitely think we need to also keep the old method, if the new method provides more consistency with the output if arp info is available then maybe that should be default and we fall back to the old code if arp info is empty?

murrant commented on app/Http/Controllers/Device/Tabs/PortsController.php:
```diff
@@ -175,24 +175,22 @@ public function findPortNeighbors(Port $port): array
-                    }
+        // IPv4 + IPv6 subnet if detailed
+        // fa-arrow-right green portlink on devicelink
+        if ($port->macLinkedPorts->isNotEmpty()) {
```
You need to change the eager loaded relationships drop ipv4Networks (if it is no longer used) and add macLinkedPorts.

laf commented:
@eskyuu Any update on this?

eskyuu commented:
@laf - Sorry I left this one for so long.  I have added the config option, added a few port related config options to the web UI, and also made it so the "ports_page_default" option works again.

murrant commented:
This feels like it is just covering up the fact that vrfs are not very well supported :D

eskyuu commented:
I don't feel like these changes are covering anything up.  The changes remove the VRF name from networks because I believe it is a fundamentally broken concept to attach a subnet to a VRF for the cases where LibreNMS uses the network table, plus the functionality that adds the VRF name to the network table was only fixed a couple of months ago.  I agree with your assessment of VRF support in NMS (I couldn't get it working as documented on IOS-XR or IOS-XE).

The main purpose of this PR is to allow the neighbour list to be optionally configured use the ARP table instead of the network table to find connected devices.  I need this functionality because the use of the network table to find neighbours results in some interfaces having 100s of incorrect devices listed on my network.

murrant commented:
If I'm understanding you right, doesn't this not work on large LAN segments with multiple switches. They would just show everything in the LAN segment as a neighbor.

murrant commented on app/Http/Controllers/Device/Tabs/PortsController.php:
```diff
@@ -283,7 +294,7 @@ private function getTabs(Device $device): array
         $tabs = [
             ['name' => __('Basic'), 'url' => 'basic'],
-            ['name' => __('Detail'), 'url' => ''],
+            ['name' => __('Detail'), 'url' => 'detail'],
```
This shouldn't be here. Otherwise the URL always has detail in it.

eskyuu replied:
I just had a closer look at this, and pushed another change to make the existing option work properly

murrant commented:
I'm still pretty nervous about this.  Could you separate out the default page change?

eskyuu commented:
done and new PR created for the detail tab change

eskyuu commented:
I found examples today of where I still get unwanted neighbours using the original ipv4networks code.  In this case it was CGNAT address space, where multiple devices connect to different carriers who use the same CGNAT subnet.

Based on this I would still prefer an option to choose which method to use for neighbour relationships, and an explicit config option also eliminates the concern about changing how the neighbours work for existing installations.

eskyuu commented:
I have reverted this back to a version where there is an explicit config option because I found instances where subnet with ARP failover still gave junk data.  The explicit config option also resolves any nervousness about changing the default behaviour.

We could still add options to do arp with subnet failover or subnet with ARP failover.
